compiler/rustc_query_system/src/dep_graph/serialized.rs

```diff
@@ -76,6 +80,9 @@ const DEP_NODE_PAD: usize = DEP_NODE_SIZE - 1;
 const DEP_NODE_WIDTH_BITS: usize = DEP_NODE_SIZE / 2;
 
 /// Data for use when recompiling the **current crate**.
+///
+/// There may be unused indices with DEP_KIND_NULL in this graph due to batch allocation of
+/// indices to threads.
```

Comment:
Is this batch allocation really necessary? Could we have a single atomic counter to assign the index and avoid the complexity?

Reply:
We need at least 1 thread local allocation to avoid races when marking nodes green. A single counter would probably result in some congestion as promoting nodes is fairly hot too.